File: techniques/modified-poisson/python/modified_poisson.py
"""Modified (robust) Poisson regression for risk ratios (Reference §7.9, §7.53).

When the outcome is BINARY (0/1) but you want the RISK RATIO instead of the
odds ratio, Zou (2004) showed that fitting a Poisson GLM with log link gives
unbiased log-RR estimates -- but with the wrong SEs (Poisson assumes
Var = mean, while Bernoulli has Var = p (1-p)).

Fix: refit the Poisson and replace the SEs with HUBER-WHITE SANDWICH SEs:

    Var_robust(beta) = (X' W X)^{-1} ( X' diag(e_i^2) X ) (X' W X)^{-1}

with e_i = y_i - mu_i. The point estimates are the Poisson MLEs; only the SEs
change.

Why use this:
  - With common outcomes (prevalence > 10%), odds ratios overstate the RR.
  - Log-binomial (Y ~ Bernoulli, log link) is the "right" model but often
    fails to converge.
  - Modified Poisson is the standard workaround in epidemiology.
"""
from __future__ import annotations

import math

import numpy as np
from scipy import stats


def modified_poisson(X, y) -> dict:
    """Poisson fit + sandwich SEs for risk-ratio estimation from binary y.

    Parameters
    ----------
    X : design matrix (intercept included).
    y : 0/1 outcomes.
    """
    X = np.asarray(X, dtype=float); y = np.asarray(y, dtype=float)
    n, p = X.shape
    # Poisson IRLS for beta (same as techniques/poisson-regression)
    beta = np.zeros(p); ll_prev = -np.inf
    for _ in range(50):
        eta = X @ beta
        mu = np.exp(np.clip(eta, -500, 500))
        w = np.clip(mu, 1e-12, None)
        z = eta + (y - mu) / w
        sw = np.sqrt(w); Xw = X * sw[:, None]; zw = z * sw
        beta_new, *_ = np.linalg.lstsq(Xw, zw, rcond=None)
        ll = float(np.sum(y * np.log(np.clip(mu, 1e-15, None)) - mu))
        if abs(ll - ll_prev) < 1e-8: break
        ll_prev = ll; beta = beta_new
    eta = X @ beta; mu = np.exp(eta)
    e = y - mu
    # Naive Poisson variance (would be wrong)
    XtWX = X.T @ (X * mu[:, None])
    A_inv = np.linalg.pinv(XtWX)
    naive_se = np.sqrt(np.diag(A_inv))
    # Sandwich (Huber-White) variance
    B = X.T @ (X * (e ** 2)[:, None])
    var_robust = A_inv @ B @ A_inv
    # HC0 -> HC1 small-sample correction
    var_robust *= n / (n - p)
    se_robust = np.sqrt(np.diag(var_robust))
    z_stats = beta / se_robust
    p_vals = 2 * stats.norm.sf(np.abs(z_stats))
    rr = np.exp(beta); zc = stats.norm.ppf(0.975)
    return {"beta": beta.tolist(),
            "naive_poisson_se": naive_se.tolist(),
            "robust_se": se_robust.tolist(),
            "z": z_stats.tolist(), "p_values": p_vals.tolist(),
            "risk_ratio": rr.tolist(),
            "rr_ci_lower": np.exp(beta - zc * se_robust).tolist(),
            "rr_ci_upper": np.exp(beta + zc * se_robust).tolist()}


def library_versions(X_no_const, y):
    import statsmodels.api as sm
    X = sm.add_constant(X_no_const)
    res = sm.GLM(y, X, family=sm.families.Poisson()).fit(cov_type="HC1")
    return {"statsmodels Poisson HC1 params": res.params.tolist(),
            "statsmodels Poisson HC1 bse": res.bse.tolist()}


if __name__ == "__main__":
    rng = np.random.default_rng(6)
    n = 500
    x1 = rng.normal(0, 1, n); x2 = rng.normal(0, 1, n)
    # Common-outcome scenario (~40% prevalence)
    eta = -0.4 + 0.4 * x1 - 0.3 * x2
    p_true = np.clip(np.exp(eta), 0, 1)        # log link on Bernoulli
    y = (rng.uniform(0, 1, n) < p_true).astype(int)
    X = np.column_stack([np.ones(n), x1, x2])

    print("=== Modified Poisson (binary y) ===")
    res = modified_poisson(X, y)
    for j, nm in enumerate(["intercept", "x1", "x2"]):
        print(f"  {nm:10s} beta = {res['beta'][j]:+.4f}")
        print(f"             naive SE = {res['naive_poisson_se'][j]:.4f}  "
              f"robust SE = {res['robust_se'][j]:.4f}")
        print(f"             RR = {res['risk_ratio'][j]:.4f} "
              f"[{res['rr_ci_lower'][j]:.3f}, {res['rr_ci_upper'][j]:.3f}]  "
              f"p = {res['p_values'][j]:.4g}")

    print("\n--- library (statsmodels.GLM Poisson, cov_type='HC1') ---")
    for k, v in library_versions(np.column_stack([x1, x2]), y).items():
        print(f"  {k}: {v}")

File: techniques/modified-poisson/r/modified_poisson.R
# Modified Poisson regression for risk ratios (Reference §7.9, §7.53)
# From-scratch IRLS + Huber-White sandwich SEs; library: sandwich::vcovHC.
# Run with:  Rscript modified_poisson.R

modified_poisson_scratch <- function(X, y) {
  X <- as.matrix(X); y <- as.numeric(y); n <- nrow(X); p <- ncol(X)
  beta <- rep(0, p); ll_prev <- -Inf
  for (it in seq_len(50)) {
    eta <- as.vector(X %*% beta); mu <- exp(pmin(pmax(eta, -500), 500))
    w <- pmax(mu, 1e-12); z <- eta + (y - mu) / w
    sw <- sqrt(w); Xw <- sw * X; zw <- sw * z
    beta_new <- as.vector(solve(crossprod(Xw), crossprod(Xw, zw)))
    ll <- sum(y * log(pmax(mu, 1e-15)) - mu)
    if (abs(ll - ll_prev) < 1e-8) break
    ll_prev <- ll; beta <- beta_new
  }
  eta <- as.vector(X %*% beta); mu <- exp(eta); e <- y - mu
  XtWX <- t(X) %*% (X * mu); A_inv <- solve(XtWX)
  naive_se <- sqrt(diag(A_inv))
  B <- t(X) %*% (X * e^2)
  var_robust <- A_inv %*% B %*% A_inv * n / (n - p)
  se_robust <- sqrt(diag(var_robust))
  list(beta = beta, naive_poisson_se = naive_se,
       robust_se = se_robust, z = beta / se_robust,
       p_values = 2 * pnorm(-abs(beta / se_robust)),
       risk_ratio = exp(beta),
       rr_ci_lower = exp(beta - qnorm(0.975) * se_robust),
       rr_ci_upper = exp(beta + qnorm(0.975) * se_robust))
}

library_demo <- function(X, y) {
  fit <- glm(y ~ X[, -1], family = poisson(link = "log"))
  cat("Naive Poisson SEs:\n"); print(summary(fit)$coefficients[, "Std. Error"])
  if (requireNamespace("sandwich", quietly = TRUE)
      && requireNamespace("lmtest", quietly = TRUE)) {
    cat("\nsandwich::vcovHC (HC1) -> robust SEs:\n")
    print(lmtest::coeftest(fit, vcov = sandwich::vcovHC(fit, type = "HC1")))
  }
}

if (sys.nframe() == 0) {
  set.seed(6); n <- 500
  x1 <- rnorm(n); x2 <- rnorm(n)
  p_true <- pmin(exp(-0.4 + 0.4 * x1 - 0.3 * x2), 1)
  y <- as.integer(runif(n) < p_true); X <- cbind(1, x1, x2)
  print(modified_poisson_scratch(X, y))
  cat("\n--- library ---\n"); library_demo(X, y)
}
